pkg/controllers/streaming/gateway_controller.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2020 the original author or authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package streaming
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/projectriff/system/pkg/apis"
+	streamingv1alpha1 "github.com/projectriff/system/pkg/apis/streaming/v1alpha1"
+	"github.com/projectriff/system/pkg/controllers"
+	"github.com/projectriff/system/pkg/refs"
+)
+
+// +kubebuilder:rbac:groups=streaming.projectriff.io,resources=gateways,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=streaming.projectriff.io,resources=gateways/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;update;patch;delete
+
+func GatewayReconciler(c controllers.Config) *controllers.ParentReconciler {
+	c.Log = c.Log.WithName("Gateway")
+
+	return &controllers.ParentReconciler{
+		Type: &streamingv1alpha1.Gateway{},
+		SubReconcilers: []controllers.SubReconciler{
+			GatewayChildServiceReconciler(c),
+			GatewayChildDeploymentReconciler(c),
+		},
+
+		Config: c,
+	}
+}
+
+func GatewayChildServiceReconciler(c controllers.Config) controllers.SubReconciler {
+	c.Log = c.Log.WithName("ChildService")
+
+	return &controllers.ChildReconciler{
```

Comment:
No need for a Setup() section here? I have yet to dig into the Parent/Child support, but how does it tell about everything that this controller is responsible? Is it in synch with all the `// +kubebuilder` annotations above?

Reply:
the Setup method is optional. Since we there is nothing extra to configure, we don't need it.

This sub-reconciler is only concerned about managing the service for the gateway.

The `// +kubebuilder` annotations are for rbac, which are manually defined for the controller.